crates/flox-rust-sdk/src/models/publish.rs

```diff
@@ -0,0 +1,117 @@
+use flox_types::catalog::CatalogEntry;
+use runix::flake_ref::git::GitRef;
+use runix::flake_ref::{protocol, FlakeRef};
+use runix::installable::AttrPath;
+use thiserror::Error;
+
+use crate::flox::Flox;
+use crate::providers::git::{GitCommandProvider as Git, GitProvider};
+
+/// State for the publish algorihm
+#[allow(dead_code)] // until we implement methods for Publish
+pub struct Publish<'flox> {
+    flox: &'flox Flox,
+    publish_ref: PublishRef,
+    attr_path: AttrPath,
+    catalog: Git,
+    analysis: Option<CatalogEntry>, // model as type state?
+}
+
+impl<'flox> Publish<'flox> {
+    pub async fn new(
+        flox: &'flox Flox,
+        publish_ref: PublishRef,
+        attr_path: AttrPath,
+    ) -> PublishResult<Publish<'flox>> {
+        let url = match publish_ref {
+            PublishRef::Ssh(ref ssh_ref) => ssh_ref.url.as_str().to_owned(),
+
+            PublishRef::Https(ref https_ref) => https_ref.url.as_str().to_owned(),
+        };
+
+        let repo_dir = tempfile::tempdir_in(&flox.temp_dir).unwrap().into_path(); // todo catch error
+        let catalog = <Git as GitProvider>::clone(url, &repo_dir, false)
+            .await
+            .unwrap(); // todo: catch error
+
+        if catalog.list_branches().await.unwrap() // todo: catch error
+            .into_iter().any(|info| info.name == "catalog")
+        {
+            catalog.checkout("catalog", false).await.unwrap(); // todo: catch error
+        } else {
+            todo!();
+            // catalog.checkout("catalog", true).await.unwrap(); // todo: catch error
+
+            // catalog.set_upstream("origin", "catalog").await.unwrap();  // todo: implement
+            //                                                               todo: catch error
+        }
+
+        Ok(Self {
+            flox,
+            publish_ref,
+            attr_path,
+            catalog,
```

Comment:
The bash impl seems to translate hydraJobs outputs into packages attrPaths?

https://github.com/flox/flox/blob/e5793cfddae8cfb6db3429696f2ac63ac4586728/flox-bash/lib/commands/publish.sh#L338

Though it also questions whether thats needed..
Lets skip that for  now

Reply:
The bash impl seems to translate hydraJobs outputs into packages attrPaths?

https://github.com/flox/flox/blob/e5793cfddae8cfb6db3429696f2ac63ac4586728/flox-bash/lib/commands/publish.sh#L338

Though it also questions whether thats needed..
Lets skip that for  now